063-topic-replication-factor-change.md

```diff
@@ -0,0 +1,221 @@
+- **Ongoing**: replicas is updated and Cruise Control has accepted the request, but the execution is not completed.
+- **Ready**: replicas is updated and Cruise Control execution is completed.
+
+When Cruise Control integration is enabled and a replicas change is detected by the BatchingTopicController, the event is notified to the ReplicasChangeManager.
+The ReplicasChangeManager is responsible for managing the communication with Cruise Control's REST API and updating the KafkaTopic status.
+
+<p align="center"><img src="images/063-topic-replication-factor-change.png"/></p>
+
+The KafkaTopic resources with replicas change are added to a pendingChanges list the first time they are notified to the ReplicasChangeManager.
+When a replicas change is marked as pending, other replicas changes for the same KafkaTopic are ignored until the pending replicas change is sent to Cruise Control and the execution is completed.
+
+The KafkaTopic resources marked as pending are added to an ongoingChanges map keyed by `User-Task-ID` when the topic configuration request is accepted by Cruise Control.
+This map is used to periodically check Cruise Control's tasks and update the KafkaTopic status.
+
+#### Configuration
+
+To request a replicas change, the Topic Operator needs to authenticate and send HTTP requests to the Cruise Control's REST API.
+The following environment variables are set automatically by the Cluster Operator when Cruise Control is configured in Kafka resource, or manually by the user in case of standalone deployment.
+
+| Name                                   | Type    | Default | Description                                                   |
+|:---------------------------------------|---------|---------|---------------------------------------------------------------|
+| STRIMZI_CRUISE_CONTROL_ENABLED         | Boolean | false   | Whether Cruise Control integration is enabled                 |
+| STRIMZI_CRUISE_CONTROL_NAMESPACE       | String  | ""      | Namespace where Cruise Control is running                     |
+| STRIMZI_CRUISE_CONTROL_HOSTNAME        | String  | ""      | Cruise Control hostname                                       |
+| STRIMZI_CRUISE_CONTROL_PORT            | Integer | 9090    | Cruise Control port                                           |
+| STRIMZI_CRUISE_CONTROL_SSL_ENABLED     | Boolean | false   | Whether SSL encryption is enabled                             |
+| STRIMZI_CRUISE_CONTROL_AUTH_ENABLED    | Boolean | false   | Whether Basic authentication is enabled                       |
+| STRIMZI_CRUISE_CONTROL_RACK_ENABLED    | Boolean | false   | Whether rack awareness is enabled in the target Kafka cluster |
+| STRIMZI_CRUISE_CONTROL_SECRET_NAME     | String  | ""      | Secret name containing the truststore                         |
+| STRIMZI_CRUISE_CONTROL_API_SECRET_NAME | String  | ""      | Secret name containing the API credentials                    |
```

Comment:
Please explain why not mount these directly (if it has any advantages). Mounting them will make a lot of things much easier:
* RBAC
* Not overloading Kube APIs
* Expecially in the standalone operator case also the structure of the secrets.

You should also explain what credentials will it use and how will they be managed.

Reply:
> Please explain why not mount these directly (if it has any advantages).

Do you mean using the env vars to pass truststore and credentials data, instead of secret names right? One problem I see is that CC credentials are generated after the TO is deployed, so I wouldn't know what to mount in the Cluster Operator deployment use case. Then, I would need 4 env vars instead of 2.

> You should also explain what credentials will it use and how will they be managed.

This is explained in the configuration section.

Reply:
@scholzj if it's fine to deploy CC first and then the EO when CC integration is enabled, then I can explore this solution.

Reply:
To be clear, it is not something I tried. Maybe I'm missing some interdependency. But it seems to me like something what should work and migth be the easiest way to address this.

Reply:
I'm adopting your suggestions here, and removing the Secret RBAC from the draft implementation. See config parameters section. Thanks.